### PR TITLE
Separate HTTP Instrumentation for Querier

### LINKED ISSUE #2707: Querier API Metrics are not reported when running in Single-Binary mode
The querier registers a separate router when running in a single binary. This router is not currently instrumented with metrics.

### PR DESCRIPTION
**Pull request by jtlisi**
**What this PR does**:

When running cortex as a single binary the requests processed by the querier are not instrumented. This PR adds a separate histogram metric `cortex_querier_request_duration_seconds` that allows the operator to make this distinction.

The metric is registered in both single binary and microservice mode. This will allow the new metric to be used for dashboards/panels in both modes. 

**Which issue(s) this PR fixes**:
Fixes #2707 

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -65,17 +65,29 @@ type API struct {
...
+	querierRequestDuration *prometheus.HistogramVec
 }
 
-func New(cfg Config, s *server.Server, logger log.Logger) (*API, error) {
+func New(cfg Config, s *server.Server, logger log.Logger, reg prometheus.Registerer) (*API, error) {
 	// Ensure the encoded path is used. Required for the rules API
 	s.HTTP.UseEncodedPath()
 
+	// Prometheus histograms for requests.
+	querierRequestDuration := prometheus.NewHistogramVec(prometheus.HistogramOpts{
```
Comment (pracucci) [inline pkg/api/api.go:77]: You can simply use `promauto.With(reg).NewHistogramVec()`.

File: pkg/api/api.go [hunk 2]
```diff
@@ -331,31 +350,34 @@ func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine
 	}))
 }
 
+// RegisterQueryAPI registers the Prometheus routes supported by the
+// Cortex querier service. Currently this can not be registered simultaneously
+// with the Querier.
+func (a *API) RegisterQueryAPI(handler http.Handler) {
```
Comment (pracucci) [inline pkg/api/api.go:356]: Is this refactoring related to the scope of the PR? I'm not against it if you need it in a project vendoring Cortex but, if it's not required, probably the previous version was more clear.

File: pkg/api/api.go (hunk 1 above)
Comment (pracucci) [inline pkg/api/api.go:77]: Have you considered moving this to `RegisterQuerier()`? The metric would be registered only if `RegisterQuerier()` is used and the change isolated there.

Comment (jtlisi) [reply pkg/api/api.go:77]: I wanted to avoid passing a `prometheus.Registerer` to that function. However, maybe it would make more sense to store the Registerer in the API struct and and initiate the metric is in the `RegisterQuerier` function.

Comment (pracucci) [reply pkg/api/api.go:77]: Yes, or you can pass the metric itself to the `RegisterQuerier()` (ie. in Thanos it's quite a common design pattern).

Comment (jtlisi) [reply pkg/api/api.go]: 👍 Cool I updated it to pass a metric as a parameter.

File: pkg/api/api.go (hunk 2 above)
Comment (pracucci) [reply pkg/api/api.go:350]: Could you comment on this? I'm fine if you need it in a project vendoring Cortex, otherwise the previous version looks more explicit to me.

**Comment by jtlisi** [comment]
> https://github.com/cortexproject/cortex/pull/2708#discussion_r440727442

I don't have any particular use case for that function. It came about when I was expirementing with alternative ways of solving the "separate querier instrumentation" problem. I kept it because I thought it looked cleaner. However, I can return it to it's previous form if you fell otherwise.

**Comment by pracucci** [comment]
Up to your judgement, but if we keep it we could make it unexported so that
we signal it's just used internally.
